### ISSUE TITLE
Add scikit-learn compatible API

### ISSUE THREAD

**Issue by Innixma** [autogluon/autogluon#1493]
Related: https://github.com/awslabs/autogluon/issues/1479

Add a scikit-learn compatible API wrapper of TabularPredictor:

- [ ] TabularClassifier
- [ ] TabularRegressor

Required functionality (may need more than listed):

- [ ] init API
- [ ] fit API
- [ ] predict API
- [ ] works in sklearn pipelines

**Comment by jckkvs** [comment]
@Innixma 
Thanks.

Here is the code.
Its main purpose is to be able to be used with sklearn's Pipeline, MultiOutputRegressor and (GridSearchCV). 

```
import numpy as np
import pandas as pd
import autogluon as ag
from autogluon.tabular  import TabularPredictor

class AutoGluonPredictor(TabularPredictor):
    def __init__(self,
                 label='',
                 problem_type=None,
                 eval_metric=None,
                 path=None,
                 verbosity=0,
                 sample_weight=None,
                 weight_evaluation=False,
                 groups=None,
                 presets='best_quality'):
        super().__init__(label=label,
                         problem_type=problem_type,
                         eval_metric=eval_metric,
                         path=path,
                         verbosity=verbosity,
                         sample_weight=sample_weight,
                         weight_evaluation=weight_evaluation,
                         groups=groups)
        
        self.presets = presets

    def fit(self, X, y=None, sample_weight=None, check_input=True):
        X_df = pd.DataFrame(X)
        y_df = pd.DataFrame(y)
        X_names = X_df.columns
        y_names = y_df.columns
        # This code has not been well considered, but is to resolve some errors. 
        if len(X_names) != len(set(X_names)):
            X_names = [f'X{i}' for i in range(len(X_names))]
        else:
            if len(list(set(X_names) & set(y_names))) >= 1 :
                X_names = [f'X{i}' for i in range(len(X_names))]
                y_names = ['y0']

        X_df.columns = X_names
        y_df.columns = y_names
        y_name = y_names[0]
        self.X_names = X_names
        self.y_names = y_names
        ###########

        self.__init__(label=y_name)
        train_data = pd.concat([X_df, y_df], axis=1)
        super().fit(train_data, presets=self.presets)    
        return self

    def predict(self, X, y=None, sample_weight=None, check_input=True):
        X_df = pd.DataFrame(X)
        X_df.columns = self.X_names

        return super().predict(X_df)

    def set_params(self, **parameters):
        for parameter, value in parameters.items():
            setattr(self, parameter, value)
        return self

    def get_params(self, deep=True):
        """
        Parameters
        ----------
        deep : Ignored. (for compatibility with sklearn)
        Returns
        ----------
        self : returns an dictionary of parameters.
        """

        params = {}
        return params        

```

I'm not thinking of using GridSearchCV for AutoGluon, but I have defined set_params to describe it in the same notation as other estimators (like Lasso). 

Also, in my special usage, I sometimes connect variable selection with Pipeline before AutoGluon, so I added special processing in fit for X_names and y_names. If you don't need it, please ignore it. 


The usage example is described below. 

```
estimator = AutoGluonPredictor()
estimator = MultiOutputRegressor(estimator)
param_grid = {}
model_name = 'AutoGluon'

from sklearn.pipeline import Pipeline
pipe = Pipeline([('selector', SelectFromModel(Lasso()) ), ('estimator', estimator)])
gsv = GridSearchCV(pipe, param_grid= {}) # No-tuning

otherpipe = Pipeline([('selector', SelectFromModel(Lasso()) ), ('estimator', Lasso())])
othergsv =  GridSearchCV(otherpipe , param_grid= {'alpha':[0.1,0.01]}) 


```

**Comment by mglowacki100** [comment]
@Innixma I'm not sure if it would be better to create separate issue, but I've noticed a few things regarding autogluon API compatibility with sklearn-API. 
Fixing those issues will allow use autogluon as drop-in replacements with following tools from sklearn ecosystem:

a) **deepchecks** (https://github.com/deepchecks/deepchecks), here is discussion https://github.com/deepchecks/deepchecks/issues/2364
but gist of issuse that autogluon lacks from sklearn API:
- `classes_` property (I think it could be provided as alias for `class_labels`)
- `feature_importances_` property (I think it could be provided as alias for `feature_importance` with some defaults, I'm not sure what is better `original` or `transformed`)

b) **cleanlab** (https://github.com/cleanlab/cleanlab)
I've tried to run it with autogluon:
```
# cleanlab works with **any classifier**. Yup, you can use sklearn/PyTorch/TensorFlow/XGBoost/etc.
cl = cleanlab.classification.CleanLearning(sklearn.YourFavoriteClassifier())

# cleanlab finds data and label issues in **any dataset**... in ONE line of code!
label_issues = cl.find_label_issues(data, labels)

# cleanlab trains a robust version of your model that works more reliably with noisy data.
cl.fit(data, labels)
```
but I've got error that estimart is not clonable by `sklearn.base.clone` (https://docs.cleanlab.ai/stable/cleanlab/classification.html)
details:
https://github.com/cleanlab/cleanlab/blob/00776646a7764e07cc4078e8c25e873ee413c5f7/cleanlab/count.py Line 997:
```
for k, (cv_train_idx, cv_holdout_idx) in enumerate(kf.split(X=labels, y=labels)):
        try:
            clf_copy = sklearn.base.clone(clf)  # fresh untrained copy of the model
        except Exception:
            raise ValueError(
                "`clf` must be clonable via: sklearn.base.clone(clf). "
                "You can either implement instance method `clf.get_params()` to produce a fresh untrained copy of this model, "
                "or you can implement the cross-validation outside of cleanlab "
                "and pass in the obtained `pred_probs` to skip cleanlab's internal cross-validation"
            )
```
I didn't look into it, but maybe it is not hard to fix, as autogluon model seems to provide get_params:
https://auto.gluon.ai/stable/_modules/autogluon/core/models/abstract/abstract_model.html#AbstractModel.get_params
unfortunately `TabularPredictor` doesn't have it.

**Comment by mglowacki100** [comment]
Hi @Innixma @jckkvs 

I've implemented draft wrappers for: 
- classifier https://colab.research.google.com/drive/1tMVNbYhu9sLsUWdq1tPabid7-eb8ZPo3?usp=sharing (autogluon with cleanlab that requires scikit-learn API)
- regressor https://colab.research.google.com/drive/1zXUTrkKmAG472VgxwCwOGyevE7BDmWgy?usp=sharing (gridSearchCV, sklearn pipeline example)

Both wrapper follow the same pattern, instead of "inheritance" from autogluon like in jckvs example, I use composition and  inheritance only from "scikit mixins".

**Classifier:**
```
class AutogluonClassifier(BaseEstimator, ClassifierMixin):
    def __init__(self, eval_metric=None, time_limit=60, hyperparameters=None, verbosity=2):
        self.eval_metric = eval_metric
        self.time_limit = time_limit
        self.hyperparameters = hyperparameters
        self.verbosity = verbosity
        self.tabular_predictor = None
        self.classes_ = None
        self.n_features_in_ = None
        self.n_classes_ = None
        self.feature_names_in_ = None

    def fit(self, X, y):
        self.n_features_in_ = X.shape[1]
        self.feature_names_in_ = X.columns.tolist()

        X = pd.DataFrame(X)
        X['target'] = y
        self.tabular_predictor = TabularPredictor(label='target', eval_metric=self.eval_metric)
        self.tabular_predictor.fit(train_data=X, time_limit=self.time_limit, hyperparameters=self.hyperparameters, verbosity=self.verbosity)

        self.classes_ = self.tabular_predictor._learner.label_cleaner.ordered_class_labels
        self.n_classes_ = len(self.classes_)

    def predict(self, X):
        if self.tabular_predictor is None:
            raise NotFittedError("Model has not been fitted. Call 'fit' before 'predict'.")
        return self.tabular_predictor.predict(X)

    def predict_proba(self, X):
        if self.tabular_predictor is None:
            raise NotFittedError("Model has not been fitted. Call 'fit' before 'predict_proba'.")
        return self.tabular_predictor.predict_proba(X)

```

**Regressor:**
```
class AutogluonRegressor(BaseEstimator, RegressorMixin):
    def __init__(self, eval_metric=None, time_limit=60, hyperparameters=None, verbosity=2):
        self.eval_metric = eval_metric
        self.time_limit = time_limit
        self.hyperparameters = hyperparameters
        self.verbosity = verbosity
        self.tabular_predictor = None

    def fit(self, X, y):
        X = pd.DataFrame(X)
        X['target'] = y
        self.tabular_predictor = TabularPredictor(problem_type='regression', label='target', eval_metric=self.eval_metric)
        self.tabular_predictor.fit(train_data=X, time_limit=self.time_limit, hyperparameters=self.hyperparameters, verbosity=self.verbosity)

    def predict(self, X):
        if self.model is None:
            raise NotFittedError("Model has not been fitted. Call 'fit' before 'predict'.")
        return self.tabular_predictor.predict(X)

```
Note that is draft version and it requires further testing/checking, also it could be extended in two ways:
- expose more autogluon "internal parameters" e.g. `sample_weight` in wrapper __init__
- add more scikit-learn methods and attributes e.g. `feature_importances_`

I think those wrappers could be added to Tutorials section (like this extension: https://auto.gluon.ai/stable/tutorials/tabular/advanced/tabular-multilabel.html) with some additional info regarding differences between autogluon and scikit-learn APIs and design consideration. Additional rationale:
- additional tutorials/documentation releases don't have to be tightly coupled with autogluon releases
- I don't think that is possible to create perfect wrapper for everybody, either some internals will be hidden or "init" will be bloated, so those wrapper are starter-kit
- with tutorial format is easier to provide some nuance, e.g. in GridSearchCV it is better to not provide `scoring` to it, but initialize autgluon wrapper with proper `eval_metric`

What do you think about it?

**Comment by Innixma** [comment]
Example of wrapper used in AutoGluon-Timeseries for Tabular Predictor: https://github.com/autogluon/autogluon/blob/4cb0fc82ecd89025a37f8c474a23896b089afb2b/timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py#L30

**Comment by Innixma** [comment]
Thanks @jckkvs and @mglowacki100 for the help here!

I've gone forward with adopting @mglowacki100's approach with some adjustments. Note that this isn't battle-tested, so it might have bugs. My implementation is available in #3769 to be available as an experimental feature in v1.0.
